titan/utils.py

```diff
@@ -225,7 +225,50 @@ def connected_components(graph):
     """
     Get connected components in graph
 
+    args:
+        graph: the model's underlying graph
+
     returns:
         list of connected components
     """
     return list(graph.subgraph(c).copy() for c in nx.connected_components(graph))
+
+
+def get_independent_bin(rand_gen, bin_def: ObjMap) -> int:
+    """
+    Get the bin key given independent bins.  A probability is selected at random, then each bin's `prob` is compared to it, the first bin that has a `prob` less than or equal to that probability is returned.
+
+    args:
+        rand_gen: A random number generator
+        bin_def: The ObjMap containing the bins
+
+    returns:
+        The integer key of the matched bin (or last bin if no matches)
+    """
+    rand_val = rand_gen.random()
+    for bin, fields in bin_def.items():
```

Comment:
I'm a little confused by this. If they're independent bins, if you have `{'a': 0.5, 'b': 0.3, 'c': 0.2}` it could never hit anything but a and likely wouldn't get a value?

Reply:
but an independent bin would usually be something like `{'1': 0.5, '2': 0.8, '3': 1.0}`, so if someone didn't have a `1.0` bin, then it would just get the last bin if nothing strictly matched `<= rand_val`

Reply:
I think my thought on "cumulative" vs "independent" was backwards from this! Understood now.